### PR TITLE
Large version bubble Support

### PR DESCRIPTION
**Pull request by A-And**
Opening this up for preliminary comments. 
The below changes pass all Pri-0 tests both with and without the `/largeversionbubble` switch enabled. Without the switch specified the changes don't introduce any functional changes.

These changes treat the entire input as a singular version bubble. 

cc @sergiy-k

### CODE REVIEW

**Comment by mjsabby** [comment]
Will there be a runtime data structure that will tell which assemblies belong to which version bubble?

**Comment by mjsabby** [comment]
Actually it should not be needed for the feature I'm thinking about. I'm thinking that for tiered compilation we should have a heuristic that can tell if it's profitable to rejit a method or not and being in the same version bubble should make it less likely that we should rejit it. Of course that in of itself maybe naive, but then the thought is if could tell MPGO that information that all of it could be done offline.

**Comment by A-And** [comment]
@mjsabby - that's a planned feature and is reasonably easy to accomplish. It very likely won't be part of this PR and the exact format specification needs to be determined.

File: src/utilcode/pedecoder.cpp [hunk 1]
```diff
@@ -2481,14 +2481,34 @@ PTR_CVOID PEDecoder::GetNativeManifestMetadata(COUNT_T *pSize) const
...
-        PRECONDITION(CheckNativeHeader());
+        PRECONDITION(HasReadyToRunHeader() || CheckNativeHeader());
         POSTCONDITION(CheckPointer(RETVAL, NULL_OK)); // TBD - may not store metadata for IJW
         NOTHROW;
         GC_NOTRIGGER;
     }
     CONTRACT_END;
+    
+    IMAGE_DATA_DIRECTORY *pDir;
+    if (HasReadyToRunHeader() && IsLargeVersionBubbleEnabled())
```
Comment (jkotas) [inline src/utilcode/pedecoder.cpp]: `IsLargeVersionBubbleEnabled` should be used during compile time only. It is not valid at runtime (always returns false). This piece of code runs at runtime. I suspect that this IsLargeVersionBubbleEnabled effectively disabled loading of large version bubble assemblies.

Comment (A-And) [reply src/utilcode/pedecoder.cpp]: Absolutely correct. Thanks for the spot.

**Comment by alpencolt** [comment]
@A-And does the `LargeVersionBubble` option mean that we have to recompile all images on updating of  any DLL (like FNV mode)?
Are there any description in documentation?

cc @gbalykov @kvochko @kbaladurin

File: src/vm/zapsig.cpp [hunk 2]
```diff
@@ -1127,7 +1127,7 @@ BOOL ZapSig::EncodeMethod(
     TypeHandle ownerType;
 
 #ifdef FEATURE_READYTORUN_COMPILER
-    if (IsReadyToRunCompilation())
+    if (IsReadyToRunCompilation() && !pMethod->GetModule()->IsInCurrentVersionBubble())
```
Comment (davidwrighton) [inline src/vm/zapsig.cpp]: > && !pMethod->GetModule()->IsInCurrentVersionBubble() [](start = 33, length = 53)

This is a meaningful behavior change even when this feature isn't enabled.

File: src/vm/zapsig.cpp [hunk 3]
```diff
@@ -1186,7 +1186,7 @@ BOOL ZapSig::EncodeMethod(
     }
 
 #ifdef FEATURE_READYTORUN_COMPILER
-    if (IsReadyToRunCompilation())
+    if (IsReadyToRunCompilation() && !pMethod->GetModule()->IsInCurrentVersionBubble())
```
Comment (davidwrighton) [inline src/vm/zapsig.cpp]: > && !pMethod->GetModule()->IsInCurrentVersionBubble( [](start = 33, length = 52)

This is a meaningful behavior change even when this feature isn't enabled.

File: src/vm/zapsig.cpp (hunk 2 above)
Comment (A-And) [reply src/vm/zapsig.cpp]: This was absolutely correct and the check was making a functional difference. Changed!

File: src/vm/zapsig.cpp [hunk 4]
```diff
@@ -1127,7 +1127,11 @@ BOOL ZapSig::EncodeMethod(
     TypeHandle ownerType;
 
 #ifdef FEATURE_READYTORUN_COMPILER
-    if (IsReadyToRunCompilation())
+
+    // For methods encoded outside of the version bubble, we use pResolvedToken which describes the metadata token from which the method originated
+    // For tokens inside the version bubble we are not constrained by the contents of pResolvedToken and as such we skip this codepath
+    // FUTURE: This condition should likely be changed or reevaluated once support for smaller version bubbles is implemented.
+    if (IsReadyToRunCompilation() && !IsLargeVersionBubbleEnabled())
```
Comment (davidwrighton) [inline src/vm/zapsig.cpp]: >IsLargeVersionBubbleEnabled [](start = 38, length = 27)

I think this still isn't right. It should be something like IsReadyToRunCompilation() && (!IsLargeVersionBubbleEnabled() ||| !pMethod-.GetModule()-.IsInCurrentVersionBubble())

File: src/vm/zapsig.cpp [hunk 5]
```diff
@@ -1186,7 +1190,9 @@ BOOL ZapSig::EncodeMethod(
     }
 
 #ifdef FEATURE_READYTORUN_COMPILER
-    if (IsReadyToRunCompilation())
+
+    // FUTURE: This condition should likely be changed or reevaluated once support for smaller version bubbles is implemented.
+    if (IsReadyToRunCompilation() && !IsLargeVersionBubbleEnabled())
```
Comment (davidwrighton) [inline src/vm/zapsig.cpp]: >IsLargeVersionBubbleEnabled [](start = 38, length = 27)

I think this still isn't right. It should be something like IsReadyToRunCompilation() && (!IsLargeVersionBubbleEnabled() ||| !pMethod-.GetModule()-.IsInCurrentVersionBubble())

File: src/vm/zapsig.cpp (hunk 4 above)
Comment (A-And) [reply src/vm/zapsig.cpp]: Makes sense - we avoid using the metadata token in the case where large version bubbles are enabled, but the method being encoded is in the same assembly. Changed!

File: src/vm/zapsig.cpp (hunk 5 above)
Comment (A-And) [reply src/vm/zapsig.cpp]: Changed.

File: src/vm/ceeload.cpp [hunk 6]
```diff
@@ -10050,14 +10050,46 @@ IMDInternalImport *Module::GetNativeAssemblyImport(BOOL loadAllowed)
...
+    // Check if Image is even R2R?
+    if (GetFile()->IsILImageReadyToRun())
+    {
+        RETURN GetFile()->GetOpenedILimage()->GetNativeMDImport(false);
+    }
+    else
+    {
+        RETURN GetFile()->GetPersistentNativeImage()->GetNativeMDImport(false);
+    }
+}
```
Comment (fadimounir) [inline src/vm/ceeload.cpp]: Why do we need this? we should just use the function right above it, and pass false. You might need to change the POSTCONDITION(CheckPointer(RETVAL)); and add NULL_OK if loadAllowed == false

Comment (A-And) [reply src/vm/ceeload.cpp]: Changed the POSTCONDITION to cater for both cases.

File: src/inc/pedecoder.h [hunk 7]
```diff
@@ -340,6 +341,10 @@ class PEDecoder
     BOOL HasReadyToRunHeader() const;
     READYTORUN_HEADER *GetReadyToRunHeader() const;
 
+#ifdef FEATURE_READYTORUN_COMPILER
+    PTR_CVOID  GetReadyToRunManifestMetadata() const;
+#endif
```
[the remark this answers is not in the document]
Comment (A-And) [reply src/inc/pedecoder.h]: Great catch! Left-over from previous iterations. Removed.

File: src/zap/zapinfo.cpp [hunk 8]
```diff
@@ -1424,12 +1424,7 @@ CORINFO_MODULE_HANDLE ZapInfo::embedModuleHandle(CORINFO_MODULE_HANDLE handle,
                                                                 void **ppIndirection)
 {
     _ASSERTE(ppIndirection != NULL);
-
-    if (IsReadyToRunCompilation())
-    {
-        _ASSERTE(!"embedModuleHandle");
-        ThrowHR(E_NOTIMPL);
-    }
+    _ASSERTE(!IsReadyToRunCompilation() || IsLargeVersionBubbleEnabled());
```
[the remark this answers is not in the document]
Comment (A-And) [reply src/zap/zapinfo.cpp]: This is definitely not correct nor is it hit at runtime. Reverted.

File: src/inc/corcompile.h [hunk 9]
```diff
@@ -1823,4 +1824,13 @@ inline bool IsReadyToRunCompilation()
 #endif
 }
 
+inline bool IsLargeVersionBubbleEnabled()
```
Comment (jkotas) [inline src/inc/corcompile.h]: This method should be ever called at crossgen time only. It should be never called at runtime.

Could you please to move the whole method under ifdef to let compiler verify it?


```
#ifdef FEATURE_READYTORUN_COMPILER
inline bool IsLargeVersionBubbleEnabled()
{
    return g_fLargeVersionBubble;
}
#endif
```

Comment (A-And) [reply src/inc/corcompile.h]: Done! (this actually dug out a piece of dead code)

**Comment by A-And** [comment]
@dotnet-bot test this please

**Comment by A-And** [comment]
@dotnet-bot test Tizen armel Cross Checked Innerloop Build and Test please

**Comment by A-And** [comment]
@dotnet-bot  test Windows_NT x86 Release Innerloop Build and Test please

**Pull request merged**